examples/moving_horizon_estimation/mhe.py

```diff
@@ -0,0 +1,174 @@
+
+    # Dynamics
+    problem_type = ProblemType.torque_driven
+
+    # Constraints
+    constraints = ()
+
+    # Path constraint
+    X_bounds = QAndQDotBounds(biorbd_model)
+    # X_bounds.min[:, 0] = X0
+    # X_bounds.max[:, 0] = X0
+
+    # Define control path constraint
+    U_bounds = Bounds([torque_min, 0.0], [torque_max, 0.0])
+
+    # Initial guesses
+    x = X0
+    u = U0
+    X_init = InitialConditions(x, interpolation_type=InterpolationType.EACH_FRAME)
+    U_init = InitialConditions(u, interpolation_type=InterpolationType.EACH_FRAME)
+    # ------------- #
+
+    return OptimalControlProgram(
+        biorbd_model,
+        problem_type,
+        number_shooting_points,
+        final_time,
+        X_init,
+        U_init,
+        X_bounds,
+        U_bounds,
+        objective_functions,
+        constraints,
+        nb_threads=1,
+    )
+
+
+def plot_true_X(q_to_plot):
+    return X_[q_to_plot, :]
+
+
+def plot_true_U(q_to_plot):
+    return U_[q_to_plot, :]
+
+
+if __name__ == "__main__":
+
+    biorbd_model_path = "./cart_pendulum.bioMod"
+    biorbd_model = biorbd.Model(biorbd_model_path)
+
+    Tf = 3  # duration of the simulation
+    X0 = np.array([0, np.pi / 2, 0, 0])
+    N = Tf * 40  # number of shooting nodes per sec
+    noise_std = 0.05  # STD of noise added to measurements
+    T_max = 2  # Max torque applied to the model
+    N_mhe = 15  # size of MHE window
+    Tf_mhe = Tf / N * N_mhe  # duration of MHE window
+
+    X_, Y_, Y_N_, U_ = run_simulation(biorbd_model, Tf, X0, T_max, N, noise_std, SHOW_PLOTS=True)
+
+    X0 = np.zeros((biorbd_model.nbQ() * 2, N_mhe))
+    U0 = np.zeros((biorbd_model.nbQ(), N_mhe - 1))
+    X_est = np.zeros((biorbd_model.nbQ() * 2, N - N_mhe))
+    T_max = 5  # Give a bit of slack on the max torque
+
+    Y_i = Y_N_[:, :, :N_mhe]
+    ocp = prepare_ocp(
+        biorbd_model_path,
+        number_shooting_points=N_mhe - 1,
+        final_time=Tf_mhe,
+        max_torque=T_max,
+        X0=X0,
+        U0=U0,
+        data_to_track=Y_i,
+    )
+    sol = ocp.solve(
+        options_ipopt={
+            "hessian_approximation": "exact",
+            "limited_memory_max_history": 10,
+            "print_level": 1,
+            "tol": 1e-2,
+            "linear_solver": "ma57",
+            "bound_frac": 1e-5,
+            "bound_push": 1e-5,
+        }
+    )
+    data_sol = Data.get_data(ocp, sol)
+    X0, U0, X_out = warm_start_mhe(data_sol)
+    X_est[:, 0] = X_out
+    t0 = time.time()
+
+    for i in range(1, N - N_mhe):
+        Y_i = Y_N_[:, :, i : i + N_mhe]
+        ocp.modify_objective_function(
+            {"type": Objective.Lagrange.MINIMIZE_MARKERS, "weight": 1000, "data_to_track": Y_i}, 0
+        )
+        sol = ocp.solve(options_ipopt={"print_level": 0, "bound_frac": 1e-5, "bound_push": 1e-5})
+        data_sol = Data.get_data(ocp, sol)
+        X0, U0, X_out = warm_start_mhe(data_sol)
+        X_est[:, i] = X_out
```

Comment:
This could be moved in the module as well. If you add `yield` at the end of the for loop, it could be very elegant!

Reply:
Do you mean yielding a list of X_out ?